book/src/developers/designs/application_passwords.md

```diff
@@ -0,0 +1,229 @@
+non-access groups that the application may still rely on. Ultimately, in this
+case the application needs to make it's own authorisation decisions to an
+extend. Kanidm can limit which users are members of the access allowed group as
+only they can bind still as an extra layer of defence)
+
+The application must bind with its api-token if it wishes to read extended user
+information. With this, only basic info limited to anonymous rights are granted.
+
+(NOTE: We can't assume these DNs are private - I did consider making these
+app=<secret key>,dc=example,dc=com, but client applications may disclose this
+basedn in ui elements).
+
+When a user authenticates the binddn of the account is set to
+spn=user,app=name,dc=example,dc=com. This difference in base DN triggers Kanidm
+to re-route the authentication to the application specific password, rather than
+the UNIX one.
+
+## Kanidm
+
+### Application Entries
+
+A new class for applications will be added. Each application will have a single
+associated group so only members of this group will be able to bind with the
+application password for the associated application.
+
+Creating a new application will not create an associated group automatically. It
+will be the administrator who will configure the association after creating the
+application and optionally a new group. It will be possible to associate
+`idm_all_persons` to an application. Removing an application will not delete the
+associated group nor its members.
+
+When users are removed from a group associated to an application all of their
+application passwords for the application will be disabled.
+
+Applications need to have service-account like properties where tokens can be
+generated for them. These are optional since an anonymous bind to kanidm and
+searching under the basedn or application base dn will continue to work.
+
+Application should have a URL reference to help admins identify where the
+application may be located or accessed.
+
+(NOTE: Future, it could be good to allow customisable instructions for users on
+where to go to use their app password?)
+
+(NOTE: To make it easier to manage tokens, it could be good to have a
+"token-capable" class type to prevent duplication this with service-accounts,
+and allows adding tokens to future types).
+
+### Accounts
+
+The user may wish to have multiple passwords per application. Each password must
+have a label to identify it, a enabled/disabled flag and a last used timestamp.
+For example:
+
+```
+  MAIL
+    iphone: abcd...
+    laptop: bcde...
+  HTTP
+    workstation: cdef...
+```
+
+Person accounts will need a new `Attribute::ApplicationPassword` that stores a
+`ValueSetApplicationPassword`. Each value in the set is a new type to manage these
+secrets and their labeling and the references to the applications.
+
+```
+struct ApplicationPAssword {
+    label: String,
+    enabled: bool,
+    last_used: Duration,
```

Comment:
Should we be storing last_used here or should this be part of #22  - I think updating this on every bind could get *really* heavy on DB writes very quickly.

Reply:
Good point. I will keep it simple for now and remove the field.